orchagent/cbf/cbfnhgorch.cpp

```diff
@@ -0,0 +1,799 @@
+#include "cbfnhgorch.h"
+#include "crmorch.h"
+#include "bulker.h"
+#include "tokenize.h"
+#include "nhgorch.h"
+#include "nhgmaporch.h"
+#include "routeorch.h"
+
+extern sai_object_id_t gSwitchId;
+
+extern NhgOrch *gNhgOrch;
+extern CbfNhgOrch *gCbfNhgOrch;
+extern CrmOrch *gCrmOrch;
+extern NhgMapOrch *gNhgMapOrch;
+extern RouteOrch *gRouteOrch;
+
+extern sai_next_hop_group_api_t* sai_next_hop_group_api;
+
+extern size_t gMaxBulkSize;
+
+CbfNhgOrch::CbfNhgOrch(DBConnector *db, string tableName) : NhgOrchCommon(db, tableName)
+{
+    SWSS_LOG_ENTER();
+}
+
+/*
+ * Purpose:     Perform the operations requested by APPL_DB users.
+ *
+ * Description: Iterate over the untreated operations list and resolve them.
+ *              The operations supported are SET and DEL.  If an operation
+ *              could not be resolved, it will either remain in the list, or be
+ *              removed, depending on the case.
+ *
+ * Params:      IN  consumer - The cosumer object.
+ *
+ * Returns:     Nothing.
+ */
+void CbfNhgOrch::doTask(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+
+    if (!gPortsOrch->allPortsReady())
+    {
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        swss::KeyOpFieldsValuesTuple t = it->second;
+
+        string index = kfvKey(t);
+        string op = kfvOp(t);
+
+        bool success;
+        const auto &cbf_nhg_it = m_syncdNextHopGroups.find(index);
+
+        if (op == SET_COMMAND)
+        {
+            string members;
+            string selection_map;
+
+            /*
+             * Get CBF group's members and selection map.
+             */
+            for (const auto &i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "members")
+                {
+                    members = fvValue(i);
+                }
+                else if (fvField(i) == "selection_map")
+                {
+                    selection_map = fvValue(i);
+                }
+            }
+
+            /*
+             * Validate the data.
+             */
+            auto p = getMembers(members);
+
+            if (!p.first)
+            {
+                SWSS_LOG_ERROR("CBF next hop group %s data is invalid.",
+                               index.c_str());
+                it = consumer.m_toSync.erase(it);
+                continue;
+            }
+
+            /*
+             * If the CBF group does not exist, create it.
+             */
+            if (cbf_nhg_it == m_syncdNextHopGroups.end())
+            {
+                /*
+                 * If we reached the NHG limit, postpone the creation.
+                 */
+                if (gRouteOrch->getNhgCount() + NhgBase::getSyncedCount() >= gRouteOrch->getMaxNhgCount())
+                {
+                    SWSS_LOG_WARN("Reached next hop group limit. Postponing creation.");
```

Comment:
There's no specific step supported - assumption is that CBF routes will only be programmed if supported

Reply:
Can we use getMaxFcVal method for this before creating CB NHG??

Reply:
If the max FC value is 0 (or generally too low) then the NHG maps won't get created, and so neither will the CBF groups.

We could make it more efficient here by instantly processing all groups, but since this can only happen by trying to use a feature that's not supported, I don't think we need to add extra code for it